File: statusbar/music.py
#!/usr/bin/env python

import _thread
import os
import re
import subprocess
import sys
import time

import common

music_program = common.MUSIC_PROGRAM

icon_fg = common.pink
icon_bg = common.black
icon_tr = "0xff"
text_fg = common.pink
text_bg = common.black
text_tr = "0xff"

icon_color = "^c" + str(icon_fg) + "^^b" + str(icon_bg) + str(icon_tr) + "^"
text_color = "^c" + str(text_fg) + "^^b" + str(text_bg) + str(text_tr) + "^"
DELAY_TIME = 1

filename = os.path.basename(__file__)
name = re.sub("\\..*", '', filename)


def get_music_title():
    cmd = "mpc current"
    result = subprocess.run(cmd, shell=True, timeout=3, stderr=subprocess.PIPE, stdout=subprocess.PIPE)
    title = result.stdout.decode('utf-8').replace('\n', '')
    title = title.replace("'", "")  # 解决一些歌曲带'的问题
    return title


def is_playing():
    cmd = "mpc status | grep -c playing"
    result = subprocess.run(cmd, shell=True, timeout=3, stderr=subprocess.PIPE, stdout=subprocess.PIPE)
    result = result.stdout.decode('utf-8').replace('\n', '')
    if int(result) > 0:
        return True
    else:
        return False


def save_cache(var, value):
    music_cache = create_cache(var)
    cmd = "echo {} > {} ".format(value, music_cache)
    subprocess.run(cmd, shell=True, timeout=3, stderr=subprocess.PIPE, stdout=subprocess.PIPE)


def get_cache(var):
    music_cache = create_cache(var)
    cmd = "cat " + music_cache
    result = subprocess.run(cmd, shell=True, timeout=3, stderr=subprocess.PIPE, stdout=subprocess.PIPE)
    value = result.stdout.decode('utf-8').replace("\n", "")
    if value == "":
        value = 0
    return value


def create_cache(var):
    cache_dir = "~/.cache/"
    music_cache = cache_dir + var
    if not os.path.exists(music_cache):
        os.system("touch " + music_cache)
    return music_cache


def update(loop=False, set_root=True):
    while True:
        icon = "ﱘ"
        text = get_music_title()
        if "" == text:
            txt = "^s{}^{} {} ".format(name, icon_color, icon)
        else:
            bar_length = 20
            music_title_pos = get_cache("music_title_pos")
            music_title_pos = int(music_title_pos)
            music_title_show_length = len(text) - bar_length
            text = text[music_title_pos:music_title_pos + bar_length]
            if is_playing():
                music_title_pos = music_title_pos + 1
            if music_title_pos > music_title_show_length:
                music_title_pos = 0
            save_cache("music_title_pos", music_title_pos)
            txt = "^s{}^{} {} {} {} ".format(name, icon_color, icon, text_color, text)
        common.write_to_file(txt + "\n", str(name))
        if not loop:
            if set_root:
                print(txt)
                # os.system("xsetroot -name '"+str(txt)+"'")
            break
        time.sleep(DELAY_TIME)


def update_thread():
    _thread.start_new_thread(update, (False, False))


def play_or_pause():
    result = os.system("mpc pause-if-playing")
    if 0 == result:
        status = "paused"
    else:
        os.system("mpc -q play")
        status = "playing"
    return status


def click(string=''):
    match string:
        case 'L':
            play_or_pause()
        case 'M':
            os.system("mpc -q stop")
            pass
        case 'R':
            os.system("xdotool keydown Super m keyup m Super")
            pass
        case 'U':
            os.system("mpc -q prev")
            pass
        case 'D':
            os.system("mpc -q next")
            pass
        case _:
            pass


def notify():
    pass


if __name__ == "__main__":
    print(sys.argv)
    if len(sys.argv) > 1:
        if sys.argv[1] == "update":
            pass
        else:
            click(sys.argv[1])
            update(set_root=False)
    else:
        update()

File: statusbar-rs/src/music.rs
use lazy_static::lazy_static;
#[cfg(test)]
use regex::Regex;

use crate::common;
use crate::common::{Button, cmd, PackageData};

const ICON_FG: &str = common::PINK;
const ICON_BG: &str = common::BLACK;
const ICON_TR: &str = "0xff";
const TEXT_FG: &str = common::PINK;
const TEXT_BG: &str = common::BLACK;
const TEXT_TR: &str = "0xff";

lazy_static! {
    static ref ICON_COLOR: String = format!("^c{}^^b{}{}^", ICON_FG, ICON_BG, ICON_TR);
    static ref TEXT_COLOR: String = format!("^c{}^^b{}{}^", TEXT_FG, TEXT_BG, TEXT_TR);

    static ref TITLE:std::sync::Mutex<MusicTitle> = std::sync::Mutex::new(MusicTitle::new("".to_string()));
}
const NAME: &str = "music";


#[cfg(test)]
#[test]
pub fn test() {
    println!("get() = {:?}", get());
    let expected_regex = Regex::new(r"\^smusic\^\^c#ff79c6\^\^b#282a360xff\^ 󰝚.+").unwrap();
    assert!(expected_regex.is_match(&get().data));
}

pub fn get() -> PackageData {
    let playing = cmd("mpc status | grep playing | wc -l");
    let playing = playing == "1";
    let title = common::cmd("mpc current");
    if title.is_empty() {
        let text = format!("^s{}^{} {} ", NAME, *ICON_COLOR, "󰝚");
        return PackageData::new(NAME, text);
    }
    let mut title_s = TITLE.lock().unwrap();

    if title_s.title == title {
        let text = format!("^s{}^{} {}{} {}", NAME, *ICON_COLOR, "󰝚", *TEXT_COLOR, title_s.get_rolling_title(playing));
        return PackageData::new(NAME, text);
    }

    title_s.title = title;
    title_s.current_pos = 0;

    let text;

    text = format!("^s{}^{} {}{} {}", NAME, *ICON_COLOR, "󰝚", *TEXT_COLOR, title_s.get_rolling_title(playing));


    PackageData::new(NAME, text)
}

pub fn api(button: Button) {
    match button {
        Button::LEFT => { cmd("mpc -q toggle"); }
        Button::RIGHT => { cmd("xdotool keydown Super m keyup m Super"); }
        Button::MIDDLE => { cmd("mpc -q stop"); }
        Button::UP => { cmd("mpc -q prev"); }
        Button::DOWN => { cmd("mpc -q next"); }
    }
}

struct MusicTitle {
    title: String,
    current_pos: usize,
}

impl MusicTitle {
    fn new(title: String) -> Self {
        MusicTitle {
            title,
            current_pos: 0,
        }
    }

    fn get_rolling_title(&mut self, playing: bool) -> String {
        let title_len = self.title.chars().count();
        if title_len <= 20 {
            let rolling_title = &self.title;
            if playing { self.current_pos = (self.current_pos + 1) % title_len; }
            rolling_title.to_string()
        } else {
            let mut char_indices = self.title.char_indices();
            let start_index = char_indices.nth(self.current_pos).map(|(i, _)| i).unwrap_or(0);
            let end_index = char_indices.nth(19).map(|(i, _)| i).unwrap_or(self.title.len());
            let rolling_title = &self.title[start_index..end_index];
            if playing { self.current_pos = (self.current_pos + 1) % (title_len - 19); }
            rolling_title.to_string()
        }
    }
}